### ISSUE TITLE
[BUG] Deadlock issue raised from backend store DB

### ISSUE THREAD

**Issue by anjsam1402** [mlflow/mlflow#4353]
## MLflow Roadmap Item

This is an MLflow Roadmap item that has been prioritized by the MLflow maintainers.

### System information
- **OS Platform and Distribution**: MLflow docker image based on 'python:3.9-slim-buster'. Deployed in k8s cluster.
- **MLflow installed from**: official pip \ pypi
- **MLflow version**: 1.16.0
- **MLflow backend store**: AWS RDS MySQL 8.0.23
- **MLflow artifact root**: AWS S3 bucket
- **Python version**: 3.9
- ** command** : `mlflow server --host 0.0.0.0  --port 5000 --default-artifact-root ${BUCKET}  --backend-store-uri mysql+pymysql://${USERNAME}:${PASSWORD}@${HOST}:${PORT}/${DATABASE}`

### Describe the problem
I have been performing load test on the Mlflow Tracking server with a multiprocessing sample code. For few of the runs the following issue was raised :

> mlflow.exceptions.MlflowException: (pymysql.err.OperationalError) (1213, 'Deadlock found when trying to get lock; try restarting transaction')
[SQL: INSERT INTO latest_metrics (`key`, value, timestamp, step, is_nan, run_uuid) VALUES (%(key)s, %(value)s, %(timestamp)s, %(step)s, %(is_nan)s, %(run_uuid)s)]
[parameters: {'key': 'rmse', 'value': 0.859125957974236, 'timestamp': 1620801352207, 'step': 0, 'is_nan': 0, 'run_uuid': 'eb6d1876d37d4706a97cec4db51937a6'}]

### Code to reproduce issue
```
import time
import multiprocessing
from multiprocessing import Pool
import os
import warnings
import sys
import pandas as pd
import numpy as np
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import ElasticNet
import mlflow
import mlflow.sklearn

mlflow.set_tracking_uri({REMOTE_TRACKING_SERVER_URI}) 
mlflow.set_experiment({EXPERIMENT_NAME})

def eval_metrics(actual, pred):
    # compute relevant metrics
    rmse = np.sqrt(mean_squared_error(actual, pred))
    mae = mean_absolute_error(actual, pred)
    r2 = r2_score(actual, pred)
    return rmse, mae, r2
    
def load_data(data_path):
    data = pd.read_csv(data_path)
    # Split the data into training and test sets. (0.75, 0.25) split.
    train, test = train_test_split(data)
    
    # The predicted column is "quality" which is a scalar from [3, 9]
    train_x = train.drop(["quality"], axis=1)
    test_x = test.drop(["quality"], axis=1)
    train_y = train[["quality"]]
    test_y = test[["quality"]]
    return train_x, train_y, test_x, test_y

def train(x, alpha=0.5, l1_ratio=0.5):
    # train a model with given parameters
    warnings.filterwarnings("ignore")
    np.random.seed(40)
    mlflow.sklearn.autolog()

    # Read the wine-quality csv file (make sure you're running this from the root of MLflow!)
    data_path = "./data/wine-quality.csv"
    train_x, train_y, test_x, test_y = load_data(data_path)
    # Useful for multiple runs (only doing one run in this sample notebook)

    with mlflow.start_run(run_name='run-infy 1000 - ' + str(x)):
        # Execute ElasticNet
        lr = ElasticNet(alpha=alpha, l1_ratio=l1_ratio, random_state=42)
        lr.fit(train_x, train_y)
        # Evaluate Metrics
        predicted_qualities = lr.predict(test_x)
        (rmse, mae, r2) = eval_metrics(test_y, predicted_qualities)

        print("##############  logging params and metrics for " + str(x) + "/" + str(500) + " ###########")
        # Log parameter, metrics, and model to MLflow
        mlflow.log_param(key="alpha", value=alpha)
        mlflow.log_param(key="l1_ratio", value=l1_ratio)
        mlflow.log_metric(key="rmse", value=rmse)
        mlflow.log_metrics({"mae": mae, "r2": r2})
        mlflow.log_artifact(data_path)
        mlflow.sklearn.log_model(lr, "model")

def multiprocessing_func(x):
    time.sleep(2)
    (x, train(x, 0.5, 0.8))

if __name__ == '__main__':
    starttime = time.time()
    pool = Pool()
    pool.map(multiprocessing_func, range(1, 1000))
    pool.close()
    print()
    print('Time taken = {} seconds'.format(time.time() - starttime))
```

### What component(s), interfaces, languages, and integrations does this bug affect?
- [X] `help wanted`: Need help from community

Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [X] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

**Comment by simonbos** [comment]
I also experienced this issue, with a similar setup:
- OS Platform and Distribution: Arch Linux
- MLflow installed from: official pip \ pypi
- MLflow version: 1.17.0
- MLflow backend store:MySQL 8.0.25 locally hosted via docker
- MLflow artifact root: local file system
- Python version: 3.9.5

I noticed that I only got a Deadlock error in `store.log_metric()` if the given metric key for a run_id is not yet in the latest_metrics table (so the first log of a new metric). 

I did some digging and was able to consistently reproduce this error:
1. Run MySQL via: `sudo docker run --name=mysql-mlflow -e MYSQL_DATABASE=mlflow -e MYSQL_USER=mlflow -e MYSQL_PASSWORD=mysecretpasswd -e MYSQL_ROOT_PASSWORD=mysecretrootpasswd -d -p 3306:3306 mysql:8.0.25`
2. Install mlflow==1.17.0 (also tested with master branch at 198c07d7) and mysqlclient (in venv).
3. Running the following script:

```python
import threading
import logging
import time

import math
from mlflow.entities import Metric
from mlflow.store.tracking.dbmodels.models import SqlMetric, SqlLatestMetric
from mlflow.store.tracking.sqlalchemy_store import SqlAlchemyStore

# setup logger
logger = logging.getLogger(__name__)
ch = logging.StreamHandler()
formatter = logging.Formatter('%(asctime)s:TID-%(thread)d:%(levelname)s: %(message)s')
ch.setFormatter(formatter)
logger.addHandler(ch)
logger.setLevel(logging.INFO)

# database configuration
db_uri = "mysql://mlflow:mysecretpasswd@127.0.0.1:3306/mlflow"
default_artifact_root = "./mlruns"
store = SqlAlchemyStore(db_uri, default_artifact_root)

# create 2 runs in 1 experiment
experiment_name = f'test_experiment_{math.floor(time.time()):.0f}'
experiment_id = store.create_experiment(name=experiment_name)
run1 = store.create_run(
    experiment_id=experiment_id,
    user_id="test_user",
    start_time=time.time(),
    tags={}
)
run2 = store.create_run(
    experiment_id=experiment_id,
    user_id="test_user",
    start_time=time.time(),
    tags={}
)

# we'll try to add metric1 to run1 and metric2 to run2 at the same time
i = math.floor(time.time())

metric1 = Metric(key=f"metric_run_{i}", value=0.1, timestamp=0, step=0)
metric2 = Metric(key=f"metric_run_{i}", value=0.2, timestamp=0, step=0)


def log_metric(run_id, metric):
    # Note: a minimal version of mlflow.store.tracking.sqlalchemy_store.SqlAlchemyStore.log_metric
    #       (without the validations)
    logger.info(f'starting run {run_id}')
    with store.ManagedSessionMaker() as session:
        logged_metric = SqlMetric(
            run_uuid=run_id,
            key=metric.key,
            value=metric.value,
            timestamp=metric.timestamp,
            step=metric.step,
            is_nan=False
        )
        session.add(logged_metric)
        latest_metric = (
            session.query(SqlLatestMetric)
                .filter(
                    SqlLatestMetric.run_uuid == logged_metric.run_uuid,
                    SqlLatestMetric.key == logged_metric.key,
                )
                .with_for_update()
                .one_or_none()
        )
        assert latest_metric is None
        logger.info("going to sleep ...")
        time.sleep(4)
        # latest_metric is None, so session.merge will perform an insert
        session.merge(
            SqlLatestMetric(
                run_uuid=logged_metric.run_uuid,
                key=logged_metric.key,
                value=logged_metric.value,
                timestamp=logged_metric.timestamp,
                step=logged_metric.step,
                is_nan=logged_metric.is_nan,
            )
        )
        logger.info("going to sleep ... a second time")
        time.sleep(4)
    logger.info(f"committed run {run_id}")


def log_metric_wrapper(run_id, metric):
    try:
        log_metric(run_id, metric)
    except Exception as e:
        logger.error(e)


thread1 = threading.Thread(target=log_metric_wrapper, args=(run1.info.run_id, metric1))
thread2 = threading.Thread(target=log_metric_wrapper, args=(run2.info.run_id, metric2))
thread1.start()
thread2.start()
```

The purpose of .with_for_update() is as commented in the code to ensure transactional isolation by setting a row-level lock (using the sql directive FOR UPDATE). However, if none of the rows adheres to the filter condition (run_uuid, metric.key), the row-level lock will be set on a non-existent row. This can have some strange behaviours as mentioned in these references:
- https://stackoverflow.com/questions/17068686/how-do-i-lock-on-an-innodb-row-that-doesnt-exist-yet
- https://fastmail.blog/advanced/mysql-lock-nonexistent-row/
- https://mysqlquicksand.wordpress.com/2019/12/20/select-for-update-on-non-existent-rows/

In our case, instead of performing row-locking, both FOR UPDATE statements probably lock the whole table and are both allowed to do so, creating a deadlock. Some more info on this deadlockfrom (`show engine innodb status;`):

```
------------------------
LATEST DETECTED DEADLOCK
------------------------
2021-06-11 12:10:26 140559797991168
*** (1) TRANSACTION:
TRANSACTION 2539, ACTIVE 8 sec inserting
mysql tables in use 1, locked 1
LOCK WAIT 6 lock struct(s), heap size 1136, 3 row lock(s), undo log entries 1
MySQL thread id 9, OS thread handle 140560126625536, query id 125 172.17.0.1 mlflow update
INSERT INTO latest_metrics (`key`, value, timestamp, step, is_nan, run_uuid) VALUES ('metric_run_1623413418', 0.2e0, 0, 0, 0, '050ec1d4e62c409ba75010228eec6605')

*** (1) HOLDS THE LOCK(S):
RECORD LOCKS space id 22 page no 4 n bits 72 index PRIMARY of table `mlflow`.`latest_metrics` trx id 2539 lock_mode X
Record lock, heap no 1 PHYSICAL RECORD: n_fields 1; compact format; info bits 0
 0: len 8; hex 73757072656d756d; asc supremum;;


*** (1) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 22 page no 4 n bits 72 index PRIMARY of table `mlflow`.`latest_metrics` trx id 2539 lock_mode X insert intention waiting
Record lock, heap no 1 PHYSICAL RECORD: n_fields 1; compact format; info bits 0
 0: len 8; hex 73757072656d756d; asc supremum;;


*** (2) TRANSACTION:
TRANSACTION 2538, ACTIVE 8 sec inserting
mysql tables in use 1, locked 1
LOCK WAIT 6 lock struct(s), heap size 1136, 3 row lock(s), undo log entries 1
MySQL thread id 8, OS thread handle 140560127969024, query id 126 172.17.0.1 mlflow update
INSERT INTO latest_metrics (`key`, value, timestamp, step, is_nan, run_uuid) VALUES ('metric_run_1623413418', 0.1e0, 0, 0, 0, 'd1318f83dbfd4e2cb13d525c314cbb66')

*** (2) HOLDS THE LOCK(S):
RECORD LOCKS space id 22 page no 4 n bits 72 index PRIMARY of table `mlflow`.`latest_metrics` trx id 2538 lock_mode X
Record lock, heap no 1 PHYSICAL RECORD: n_fields 1; compact format; info bits 0
 0: len 8; hex 73757072656d756d; asc supremum;;


*** (2) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 22 page no 4 n bits 72 index PRIMARY of table `mlflow`.`latest_metrics` trx id 2538 lock_mode X insert intention waiting
Record lock, heap no 1 PHYSICAL RECORD: n_fields 1; compact format; info bits 0
 0: len 8; hex 73757072656d756d; asc supremum;;

*** WE ROLL BACK TRANSACTION (2)
```

There are multiple options mentioned in the references for 'solving' this behavior, but I would need some input from maintainers on which one to choose.

**Comment by simonbos** [comment]
@anjsam1402 Could you assign a triaging label (https://github.com/mlflow/mlflow/blob/master/ISSUE_TRIAGE.rst), e.g. `help wanted` or `needs committer feedback` to get this issue back on the radar?

**Comment by simonbos** [comment]
@harupy Could you help in getting this issue back on track?

**Comment by NieuweNils** [comment]
Hi @anjsam1402 !
I'm relatively new here, but this looks interesting!
I'll take a look and see if I can help somehow.
Keep me posted if there are any updates!

**Comment by NieuweNils** [comment]
@dbczumar I just noticed it's assigned to you.
Any update on this yet?

**Comment by dbczumar** [comment]
Hi @NieuweNils ! We've identified a similar issue and fixed it in the Databricks-hosted MLflow Tracking service. I'm going to apply a corresponding fix to OSS to close out this issue shortly!

**Comment by jeremyjordan** [comment]
Hi @dbczumar, did you have a chance to apply the deadlock fix for the OSS mlflow tracking service? We've been seeing this issue and would love to get it resolved :)

**Comment by dbczumar** [comment]
Addressed via #4299. Apologies for the delay!
